prow/cmd/tide/main.go

```diff
@@ -42,47 +39,51 @@ import (
 type options struct {
 	port int
 
-	dryRun  bool
-	runOnce bool
-	deckURL string
-
 	configPath    string
 	jobConfigPath string
-	cluster       string
 
 	syncThrottle   int
 	statusThrottle int
 
-	githubEndpoint  flagutil.Strings
-	githubTokenFile string
+	dryRun     bool
+	runOnce    bool
+	kubernetes prowcmdutil.KubernetesOptions
+	github     prowcmdutil.GitHubOptions
 }
 
-func gatherOptions() options {
-	o := options{
-		githubEndpoint: flagutil.NewStrings("https://api.github.com"),
+func (o *options) Validate() error {
+	for _, group := range []cmdutil.OptionGroup{&o.kubernetes, &o.github} {
+		if err := group.Validate(o.dryRun); err != nil {
+			return err
+		}
 	}
-	flag.IntVar(&o.port, "port", 8888, "Port to listen on.")
 
-	flag.BoolVar(&o.dryRun, "dry-run", true, "Whether to mutate any real-world state.")
-	flag.BoolVar(&o.runOnce, "run-once", false, "If true, run only once then quit.")
-	flag.StringVar(&o.deckURL, "deck-url", "", "Deck URL for read-only access to the cluster.")
-
-	flag.StringVar(&o.configPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
-	flag.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
-	flag.StringVar(&o.cluster, "cluster", "", "Path to kube.Cluster YAML file. If empty, uses the local cluster.")
-
-	flag.IntVar(&o.syncThrottle, "sync-hourly-tokens", 800, "The maximum number of tokens per hour to be used by the sync controller.")
-	flag.IntVar(&o.statusThrottle, "status-hourly-tokens", 400, "The maximum number of tokens per hour to be used by the status controller.")
+	return nil
+}
 
-	flag.Var(&o.githubEndpoint, "github-endpoint", "GitHub's API endpoint.")
-	flag.StringVar(&o.githubTokenFile, "github-token-file", "/etc/github/oauth", "Path to the file containing the GitHub OAuth token.")
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.IntVar(&o.port, "port", 8888, "Port to listen on.")
+	fs.StringVar(&o.configPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
+	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
+	fs.BoolVar(&o.dryRun, "dry-run", true, "Whether to mutate any real-world state.")
+	fs.BoolVar(&o.runOnce, "run-once", false, "If true, run only once then quit.")
+	for _, group := range []cmdutil.OptionGroup{&o.kubernetes, &o.github} {
+		group.AddFlags(fs)
+	}
+	fs.IntVar(&o.syncThrottle, "sync-hourly-tokens", 800, "The maximum number of tokens per hour to be used by the sync controller.")
+	fs.IntVar(&o.statusThrottle, "status-hourly-tokens", 400, "The maximum number of tokens per hour to be used by the status controller.")
 
-	flag.Parse()
+	fs.Parse(os.Args[1:])
 	return o
 }
 
 func main() {
 	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
 
 	logrus.SetFormatter(
 		logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "tide"}),
```

Comment:
Please configure the logger before gathering options so that any error messages are properly formatted.

Reply:
> Please configure the logger before gathering options so that any error messages are properly formatted.

This seems like a fairly widespread issue.  On master:

```console
$ git describe --always --abbrev=0 --dirty
b0a96772e009a594eab5abb51c6d113f107374c9
$ git grep -nA10 '= gatherOptions()' | grep SetFormatter
prow/cmd/checkconfig/main.go-130-	logrus.SetFormatter(
prow/cmd/crier/main.go-147-	logrus.SetFormatter(
prow/cmd/deck/main.go-121-	logrus.SetFormatter(
prow/cmd/hook/main.go-103-	logrus.SetFormatter(logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "hook"}))
prow/cmd/horologium/main.go-48-	logrus.SetFormatter(
prow/cmd/jenkins-operator/main.go-122-	logrus.SetFormatter(
prow/cmd/plank/main.go-92-	logrus.SetFormatter(
prow/cmd/sinker/main.go-62-	logrus.SetFormatter(
prow/cmd/tide/main.go-87-	logrus.SetFormatter(
prow/cmd/tot/main.go-276-	logrus.SetFormatter(
```

Can I punt it to follow-up work?